### DISCUSSION TITLE
How to include resources in a product?

### DISCUSSION THREAD

**Discussion by HannesWell** [eclipse-tycho/tycho#1486] [Q&A] [answered]
Recently I wanted to add a few resources to the root of a PDE-product build with Tycho.
- a custom README file, which should be included in the linux and windows variant 
- a custom install.sh script, which should only be included in the linux variant
- 
I didn't found a elegant solution using Tycho and ended up using an extra antrun-plugin execution to copy the file after the product has been materialized and before it is packed as zip:
```
<plugin>
	<groupId>org.apache.maven.plugins</groupId>
	<artifactId>maven-antrun-plugin</artifactId>
	<version>3.1.0</version>
	<executions>
		<execution>
			<id>copy-product-resources</id>
			<goals>
				<goal>run</goal>
			</goals>
			<phase>prepare-package</phase>
			<configuration>
				<target>
					<copy file="${project.basedir}/README.txt" tofile="${project.basedir}/target/products/<product-id>/linux/gtk/x86_64/<rootFolder>/README.txt" />
					<copy file="${project.basedir}/README.txt" tofile="${project.basedir}/target/products/<product-id>/win32/win32/x86_64/<rootFolder>/README.txt" />

					<chmod file="${project.basedir}/install.sh" perm="+x" />
					<copy file="${project.basedir}/install.sh" tofile="${project.basedir}/target/products/<product-id>/linux/gtk/x86_64/<rootFolder>/install.sh" />
				</target>
			</configuration>
		</execution>
	</executions>
</plugin>
```

As you can see the copy task has to deal with paths whose segments are are mostly implementation/configuration depended and have to be kept in sync if other configuration elements change, for example the uid or the root-folder.
Furthermore the task has to be repeated for each environment.

Does anybody know a better existing solution or should we add one?

If there is non we could consider to add means to include arbitrary resources into the product, that are for example processed to the `tycho-p2-director-plugin:materialize-products-mojo` goal.
In general it should be possible to
- define the resource to include from the product's project (or at least from the containing folder)
- define a target location as relative-path from the product's root
- define a filter to only include the resource to the product for selected environments
- modify/set linux file permissions, but that's is maybe out of the scope

My first idea was to use the`build.properties` for that, but I unsure if a target-path or a filter can be specified in the usual way here.
In general something like `product.includes = README.txt;target="legal";filter=(& (osgi.os=linux) (osgi.ws=gtk) (osgi.arch=x86_64))` but that is more OSGi/MANIFEST.MF syntax and I don't know if it is a good idea to use that in a build.properties file. On the other hand I don't think it is forbidden.

**Comment by laeubi** [comment, marked as answer]
I think root files are the way of PDE to express this (if I understand your case correctly):
https://help.eclipse.org/latest/index.jsp?topic=%2Forg.eclipse.pde.doc.user%2Ftasks%2Fpde_rootfiles.htm


> modify/set linux file permissions, but that's is maybe out of the scope

you could use a p2.inf (search for `chmod`)

https://wiki.eclipse.org/Equinox/p2/Customizing_Metadata

**Reply by HannesWell** [reply]
> I think root files are the way of PDE to express this (if I understand your case correctly): https://help.eclipse.org/latest/index.jsp?topic=%2Forg.eclipse.pde.doc.user%2Ftasks%2Fpde_rootfiles.htm
> 
Thanks a lot! That's it and also works in Tycho (I have not tested PDE build, but since the doc is for PDE-build I expect that it will work too).

For the record and because the doc was not absolutely clear about that, one has to specify the root files in the `build.properties` of the feature-project that contains them. For my case described above, the following `build.properties` did exactly what I want:
```
bin.includes = feature.xml
root=file:README.txt
root.linux.gtk.x86_64=file:install.sh
root.linux.gtk.x86_64.permissions.755=install.sh
```
This way even the file-permissions are set and the `p2.inf` is not necessary.

**Reply by dhendriks** [reply]
We use this feature in a similar way in Eclipse ESCET, see [here](https://gitlab.eclipse.org/eclipse/escet/escet/-/blob/00519d5913ead7366f44a7cfaaf1782722775d00/product/org.eclipse.escet.product.headless.feature/build.properties). It works, but only if you build on Linux. If you build on Windows your Linux product will include the files but not with the right permissions.

**Reply by laeubi** [reply]
That's why I mentioned p2.inf as this is executed at install time and you can use anything the target os supports. But maybe it is fine here anyways to require run on linux.

**Reply by HannesWell** [reply]
> That's why I mentioned p2.inf as this is executed at install time and you can use anything the target os supports. But maybe it is fine here anyways to require run on linux.

Our Jenkins build server indeed runs on Linux, like probably the most server, so that is not a problem in my case.

However I wonder how the p2.inf can set the permissions for a linux-targeted build that is created e.g. on Windows? Because all the ways in java I know to set the permissions don't work on Windows, like Files.setPosixFilePermissions().
Is that because the actions defined in the p2.inf are executed on the target-mashine during installation, while the root-file approach sets them during build?

**Reply by dhendriks** [reply]
> Is that because the actions defined in the p2.inf are executed on the target-mashine during installation, while the root-file approach sets them during build?

Indeed.

**Reply by laeubi** [reply]
p2.inf would only work if you install that afterwards (e.g. via an update site) if you assemble the product as art of the build I think it makes no difference.
